sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/AlterTableAddPartitionSuite.scala

```diff
@@ -129,7 +129,9 @@ class AlterTableAddPartitionSuite
     withNamespaceAndTable("ns", "tbl") { t =>
       sql(s"CREATE TABLE $t (c int) $defaultUsing PARTITIONED BY (p int)")
 
-      withSQLConf(SQLConf.SKIP_TYPE_VALIDATION_ON_ALTER_PARTITION.key -> "true") {
+      withSQLConf(
+          SQLConf.SKIP_TYPE_VALIDATION_ON_ALTER_PARTITION.key -> "true",
+          SQLConf.ANSI_ENABLED.key -> "false") {
```

Comment:
what's the behavior of old Spark versions when ansi mode is on? @ulysses-you

Reply:
It passes if the validation is skipped. DSv2 it never worked with/without ANSI (but the exception message was different).

Reply:
I confirmed it, previous it would fail and the reason is same with this pr. We need explicitly specify the ansi mode with on/off to test..

Reply:
**Before:**


DSv1

| `spark.sql.ansi.enabled` | |
| --- | --- |
| `false` | Casted |
| `true` | Casted |

DSv2

| `spark.sql.ansi.enabled` | |
| --- | --- |
| `false` | Casted |
| `true` | Error |

**After:**

DSv1

| `spark.sql.ansi.enabled` | `spark.sql.legacy.skipTypeValidationOnAlterPartition` | |
| --- | --- | --- |
| `false` | `true` | Casted |
| `true` | `true` | Casted |
| `false` | `false` | follows `spark.sql.storeAssignmentPolicy` |
| `true` | `false` | follows `spark.sql.storeAssignmentPolicy` |

DSv2

| `spark.sql.ansi.enabled` | `spark.sql.legacy.skipTypeValidationOnAlterPartition` | |
| --- | --- | --- |
| `false` | `true` | Casted |
| `true` | `true` | Error |
| `false` | `false` | follows `spark.sql.storeAssignmentPolicy` |
| `true` | `false` | follows `spark.sql.storeAssignmentPolicy` |

Reply:
cc @ulysses-you to confirm ^

Reply:
@HyukjinKwon yeah, this is a clear behavior matrix. thank you ! the behavior is a bit complex though ..

Reply:
one nit is, before DSv1 would not do cast, just use original partition spec

Reply:
@MaxGekk this pr aims to recover ANSI CI eagerly. I will do a followup to improve the test converage. It's my bad..